### PR TITLE
feat: add wheel method to Mouse class

### LINKED ISSUE #1679: [Feature request] Mouse wheel API calls
Sounds a bit old in our age of touch interfaces, but would be great to have native mouse wheel simulation (to use with existing code).

### PR DESCRIPTION
**Pull request by christian-bromann**
fixes #1679

This patch adds a wheel command to the [`Mouse`](https://pptr.dev/#?product=Puppeteer&version=v4.0.1&show=api-class-mouse) class. It behaves similar to other mouse commands like `up` or `down` and therefor requires a `mouse.move(x, y)` for targeting the element properly.

An example script to test this feature:

```js
const puppeteer = require('./');

function sleep (ms = 3000) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

(async () => {
  const browser = await puppeteer.launch({
    headless: false,
  });
  const page = await browser.newPage();

  page.on('request', async (request) => {
    console.log(await request.headers());
  });

  await page.goto('https://mdn.mozillademos.org/en-US/docs/Web/API/Element/wheel_event$samples/Scaling_an_element_via_the_wheel?revision=1587366');
  await page.mouse.wheel({ y: 100 });
  await sleep(1000);

  const elem = await page.$('div');
  const boundingBox = await elem.boundingBox();
  console.log(boundingBox); // returns: { x: 342.5, y: 242.5, width: 115, height: 115 }

  await page.mouse.move(
    boundingBox.x + boundingBox.width / 2,
    boundingBox.y + boundingBox.height / 2
  );

  await page.mouse.wheel({ deltaY: -100 });
  console.log(await elem.boundingBox()); // returns: { x: 285, y: 185, width: 230, height: 230 }

  await sleep();
  await browser.close();
})();
```

### CODE REVIEW

File: test/assets/input/wheel.html [hunk 1]
```diff
@@ -0,0 +1,43 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <style type="text/css">
```
Comment (mathiasbynens) [inline test/assets/input/wheel.html]:
```suggestion
    <style>
```

Comment (mathiasbynens) [reply test/assets/input/wheel.html]: https://mathiasbynens.be/notes/html5-levels#level-1

File: src/common/Input.ts [hunk 2]
```diff
@@ -440,6 +445,38 @@ export class Mouse {
...
+   * await page.goto('https://mdn.mozillademos.org/en-US/docs/Web/API/Element/wheel_event$samples/Scaling_an_element_via_the_wheel?revision=1587366');
+   *
+   * const elem = await page.$('div');
+   * const boundingBox = await elem.boundingBox();
+   * await page.mouse.move(
+   *   boundingBox.x + boundingBox.width / 2,
+   *   boundingBox.y + boundingBox.height / 2
+   * );
+   *
+   * await page.mouse.wheel({ y: -100 })
```
Comment (mathiasbynens) [inline src/common/Input.ts]: `deltaY`?

**Review by mathiasbynens — CHANGES_REQUESTED** [review]
Thanks for taking this on, Christian! Looking good so far.

Please document this in `docs/api.md` as well.

File: src/common/Input.ts [hunk 3]
```diff
@@ -283,6 +283,11 @@ interface MouseOptions {
   clickCount?: number;
 }
 
+interface MouseWheelOptions {
```
Comment (jackfranklin) [inline src/common/Input.ts]: you'll need an `export` on this (so that when we generate the `puppeteer.d.ts` file it gets included)

Comment (christian-bromann) [reply src/common/Input.ts]: I assume we want to export `MouseOptions` then too.

**Comment by christian-bromann** [comment]
@mathiasbynens @jackfranklin thanks for the comments. I addressed them and squashed all commits into one. Please let me know if anything is missing.

**Review by jackfranklin — APPROVED** [review]
Thanks for this! Sorry you have to write the docs in two places; hopefully we can finish the migration in the not-too-distant future.

**Comment by christian-bromann** [comment]
> hopefully we can finish the migration in the not-too-distant future.

Sounds good! Let me know if there is anything I can do to help.

**Comment by christian-bromann** [comment]
@jackfranklin I need some help understanding why the docs test are failing:

```
DocLint Failures:
  1) [MarkDown] Method Mouse.wheel() options Object != MouseWheelOptions
  2) [MarkDown] Method Mouse.wheel() options has unexpected property 
```

They seem to scan the `api.md` file and complain about my change there:

```diff
+ #### mouse.wheel([options])
+ - `options` <[Object]>
+   - `deltaX` X delta in CSS pixels for mouse wheel event (default: 0).
+   - `deltaY` Y delta in CSS pixels for mouse wheel event (default: 0).
+ - returns: <[Promise]>
+ 
+ Dispatches a `mousewheel` event.
```

However I don't see any other TS interface options being used anywhere else in that document, e.g. for `mouse.up()`:

```md
#### mouse.up([options])
- `options` <[Object]>
  - `button` <"left"|"right"|"middle"> Defaults to `left`.
  - `clickCount` <[number]> defaults to 1. See [UIEvent.detail].
- returns: <[Promise]>

Dispatches a `mouseup` event.
```

Any advice?

**Comment by jackfranklin** [comment]
@christian-bromann ah, this is because the old system that's used to validate the docs doesn't understand the TypeScript source code as well as it should. Given we are migrating away from it I've not invested time into fixing it.

So what we do is we basically tell the old doc tool to ignore it, because it doesn't understand. You can see examples of this here: https://github.com/puppeteer/puppeteer/blob/main/utils/doclint/check_public_api/index.js#L746

This is not good, and another reason we're keen to get #6118 done!

**Review by OrKoN — APPROVED** [review]

File: docs/api.md [hunk 4]
```diff
@@ -2579,6 +2580,14 @@ Dispatches a `mousemove` event.
 
 Dispatches a `mouseup` event.
 
+#### mouse.wheel([options])
+- `options` <[Object]>
+  - `deltaX` X delta in CSS pixels for mouse wheel event (default: 0).
+  - `deltaY` Y delta in CSS pixels for mouse wheel event (default: 0).
```
Comment (mathiasbynens) [inline docs/api.md]: Please include an explanation of what positive and negative means in this context.

File: docs/api.md [hunk 5]
```diff
@@ -2579,6 +2580,14 @@ Dispatches a `mousemove` event.
...
+- returns: <[Promise]>
+
+Dispatches a `mousewheel` event.
```
Comment (mathiasbynens) [inline docs/api.md:2589]: Let's include the code example here as well

**Review by hanselfmu — APPROVED** [review]

**Review by mathiasbynens — APPROVED** [review]

**Pull request merged**